lib/Dialect/HW/HWOps.cpp

```diff
@@ -1720,6 +1720,23 @@ void ArrayGetOp::build(OpBuilder &builder, OperationState &result, Value input,
   build(builder, result, resultType, input, index);
 }
 
+// An array_get of an array_create with a constant index can just be the
+// array_create operand at the constant index.
+OpFoldResult ArrayGetOp::fold(ArrayRef<Attribute> operands) {
+  auto inputCreate = dyn_cast_or_null<ArrayCreateOp>(input().getDefiningOp());
+  if (!inputCreate)
+    return nullptr;
+
+  auto constIdx = dyn_cast_or_null<ConstantOp>(index().getDefiningOp());
+  if (!constIdx || constIdx.value().getBitWidth() > 64)
+    return nullptr;
+
+  uint64_t idx = constIdx.value().getLimitedValue();
+  auto createInputs = inputCreate.inputs();
+  assert(idx < createInputs.size());
```

Comment:
I think it is a bug of frontend when we have out-of-bounds behavior at this point, so I feel it might be better to emit warning and return nullptr as a folding result, rather than just assertion failures.

Reply:
I was thinking the verifier would catch this and throw an error by now. But the folder can be run at any point and (presumably) on invalid IR so it shouldn't just crash.

On 7.6.4: we don't want semantics of SV to be the default in the HW dialect. That said, I do think this is reasonable behavior. OTOH, I don't want to have that discussion in this PR.

Reply:
IR should never be invalid.  Either don't fold in this case or fold and let the verifier for the op catch this later.

Reply:
It's my understanding that IR is allowed to be invalid during transformation passes.

Reply:
Yes, that was too strong a statement.  This is the type of thing that undef is used for in LLVM.  The IR was well formed, but had undefined behavior.  It seems that always doing the fold, even for out-of-bounds cases isn't changing this any.